### PR TITLE
Array arguments

### PR DESCRIPTION
**Pull request by meisonlikesicecream**
Updates to allow arrays as HLS module function arguments (and some bitwidth updates). The "indices" of the ports created goes from 0 to n, except for the fullmatch where the index depends on the seed, in accordance with: https://github.com/cms-L1TK/firmware-hls/blob/master/TrackletAlgorithm/MatchCalculator.h#L814-L863

### CODE REVIEW

File: WriteVHDLSyntax.py
Comment (djcranshaw) [inline WriteVHDLSyntax.py:131]: If this line is changing, we should also change L117 and L212, correct?

File: WriteHDLUtils.py
Comment (djcranshaw) [inline WriteHDLUtils.py:839]: Rather than remove the break, could we conditionally break here if argname_is_array is false? Then we don't loop through the remaining memories looking for a port name that we have already found. Might save some run time, not sure how significant.

File: WriteHDLUtils.py
Comment (djcranshaw) [inline WriteHDLUtils.py:846]: Does this eliminate the need to know in advance the length of the array of memories that is passed to the templated function? That's great if so.

**Review by djcranshaw — CHANGES_REQUESTED** [review]
Looks good, I have just a few questions, see inline. Also, does the order of the merge w.r.t @tomalin PR matter?

**Comment by meisonlikesicecream** [comment]
I'm don't think the order of the PRs should matter, but Ian was the one who tried merging our branches (in meiiantest), so he might have a better answer.

**Comment by tomalin** [comment]
> I'm don't think the order of the PRs should matter, but Ian was the one who tried merging our branches (in meiiantest), so he might have a better answer.
> 
> EDIT: I might have lied, the changing from "(7 downto 0)" to "(6 downto 0)" was done after our test-merging, so the reason I didn't change L117 and L212 was because they didn't exist in Ians branch.

IAN: when testing, we merged Mei-Li's branch into mine. So it's probably safest to merge mine into master first.

File: WriteVHDLSyntax.py
Comment (meisonlikesicecream) [reply WriteVHDLSyntax.py:131]: Ah, I completely missed that. I didn't change them because those lines don't exist in Ians PR. Do I still change them or wait until Ians branch gets merged and just ignore it...? (I don't know what the best git procedures are... :-) )

File: WriteHDLUtils.py
Comment (meisonlikesicecream) [reply WriteHDLUtils.py:839]: Good idea! I'll add that.

File: WriteHDLUtils.py
Comment (meisonlikesicecream) [reply WriteHDLUtils.py:846]: Not sure I understand this... It doesn't check the length of the array in the actual HLS, instead it just assumes that the number of "array elements" that ends up having ports are the same as in the wiring. 

In the Matchcalculator, the size of the array for the fullmatch memories is set to 8, but only 2 memories are used so the HLS ends up only creating those two ports. So looking at the array length wouldn't really work anyway?
https://github.com/cms-L1TK/firmware-hls/blob/master/TrackletAlgorithm/MatchCalculatorTop.h#L7

Comment (djcranshaw) [reply WriteHDLUtils.py:846]: So it sounds like there never was a need to know in advance how long the memory arrays are. I had this vague worry that it was a problem that needed solving, but it seems not. This is good news anyways.

File: WriteVHDLSyntax.py
Comment (djcranshaw) [reply WriteVHDLSyntax.py:131]: Okay now I understand, those ports are no longer needed because they correspond to nentries coming out of HLS modules and going into the intermediate memories. So the best thing to do is ignore it - if you change it now there will be merge conflicts when we try to merge Ian's PR.

**Review by djcranshaw — APPROVED** [review]

**Pull request merged**
